packages/core/src/workflows/workflow.ts

```diff
@@ -62,6 +62,7 @@ export class Workflow<
   #stepGraph: StepGraph = { initial: [] };
   #stepSubscriberGraph: Record<string, StepGraph> = {};
   #steps: Record<string, IAction<any, any, any, any>> = {};
+  #onStepTransition: Array<(state: WorkflowRunState) => void | Promise<void>> = [];
```

Comment:
Move to Set data structure
```suggestion
  #onStepTransition: Set<(state: WorkflowRunState) => void | Promise<void>> = new Set();
```